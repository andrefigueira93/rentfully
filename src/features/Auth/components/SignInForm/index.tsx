/**
 * Sign In Form component.
 *
 * This component renders a form for signing in with email and password.
 * It uses react-hook-form for form validation and submission.
 *
 * @returns The Sign In Form component.
 */
import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import { useAuth } from "@/hooks/useAuth";
import { SignInProps, SignInSchema } from "@/schemas/auth";
import { zodResolver } from "@hookform/resolvers/zod";
import { ExclamationTriangleIcon } from "@radix-ui/react-icons";
import { useMutation } from "@tanstack/react-query";
import React from "react";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { toast } from "sonner";

const SignInForm: React.FC = () => {
  const { signIn } = useAuth();
  const navigate = useNavigate();

  const form = useForm<SignInProps>({
    resolver: zodResolver(SignInSchema),
    defaultValues: {
      email: "",
      password: "",
    },
    mode: "all",
    reValidateMode: "onChange",
  });

  const { mutate, isPending } = useMutation<SignInProps>({
    mutationFn: async () => {
      signIn({
        email: form.getValues("email"),
        password: form.getValues("password"),
      });
      return form.getValues();
    },
    onSuccess: () => {
      navigate("/");
    },
    onError: (error) => {
      console.error(error);
      toast("Whoops!", {
        description: "Verify your credentials",
        icon: <ExclamationTriangleIcon />,
        cancel: {
          label: "Close",
        },
        position: "top-right",
      });
    },
  });

  return (
    <Form {...form}>
      <form
        onSubmit={form.handleSubmit(() => {
          mutate();
        })}
        className="space-y-8"
      >
        <FormField
          control={form.control}
          name="email"
          render={({ field }) => (
            <FormItem>
              <FormLabel>E-mail</FormLabel>
              <FormControl>
                <Input testId="email-input" placeholder="E-mail" {...field} />
              </FormControl>
              <FormDescription>Your best email</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="password"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Password</FormLabel>
              <FormControl>
                <Input
                  testId="password-input"
                  placeholder="Password"
                  type="password"
                  {...field}
                />
              </FormControl>
              <FormDescription>Your best password</FormDescription>
              <FormMessage />
            </FormItem>
          )}
        />
        <Button
          data-testid="sign-in-button"
          disabled={isPending}
          type="submit"
          className="mt-2 w-full"
        >
          {isPending ? "Loading..." : "Sign In"}
        </Button>
      </form>
    </Form>
  );
};

export default SignInForm;
