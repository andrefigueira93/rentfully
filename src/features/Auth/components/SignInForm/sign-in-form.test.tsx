/* eslint-disable @typescript-eslint/no-explicit-any */
import { baseUser } from "@/__tests__/test-constants";
import TestWrapper from "@/__tests__/wrapper";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import SignInForm from "./index";

vi.mock("sonner");
vi.mock("@/hooks/useAuth", async () => {
  const actualModule = await vi.importActual("@/hooks/useAuth");
  return {
    ...(actualModule as any),
    useAuth: () => ({
      signIn: vi.fn(),
    }),
  };
});
vi.mock("@tanstack/react-query", async () => {
  const actualModule = await vi.importActual("@tanstack/react-query");
  return {
    ...(actualModule as any),
    useMutation: () => ({
      mutate: vi.fn(),
      isPending: false,
    }),
  };
});
vi.mock("react-router-dom", () => ({
  useNavigate: vi.fn(),
}));

describe("SignInForm", () => {
  test("renders form fields", () => {
    const { container } = render(<SignInForm />, { wrapper: TestWrapper });
    expect(screen.getByTestId("email-input")).toBeInTheDocument();
    expect(screen.getByTestId("password-input")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("handles form submission with correct credentials", async () => {
    render(<SignInForm />, { wrapper: TestWrapper });

    fireEvent.change(screen.getByPlaceholderText("E-mail"), {
      target: { value: baseUser.email },
    });
    fireEvent.change(screen.getByPlaceholderText("Password"), {
      target: { value: baseUser.password },
    });

    fireEvent.click(screen.getByText("Sign In"));

    // TODO: Finish this test
    await waitFor(() => {
      // expect(mockSignIn).toHaveBeenCalled();
      // expect(mockNavigate).toHaveBeenCalledWith('/');
    });
  });

  test("shows error on invalid submit", async () => {
    render(<SignInForm />, { wrapper: TestWrapper });
    fireEvent.click(screen.getByTestId("sign-in-button"));

    await waitFor(() => {
      expect(screen.getAllByTestId("form-message")).toHaveLength(2);
    });
  });
});
