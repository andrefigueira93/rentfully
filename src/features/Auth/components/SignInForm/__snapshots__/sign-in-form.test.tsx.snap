// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignInForm > renders form fields 1`] = `
<div>
  <form
    class="space-y-8"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        E-mail
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="email-input"
        id=":r0:-form-item"
        name="email"
        placeholder="E-mail"
        value=""
      />
      <p
        class="text-[0.8rem] text-muted-foreground"
        id=":r0:-form-item-description"
      >
        Your best email
      </p>
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Password
      </label>
      <input
        aria-describedby=":r1:-form-item-description"
        aria-invalid="false"
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        data-testid="password-input"
        id=":r1:-form-item"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <p
        class="text-[0.8rem] text-muted-foreground"
        id=":r1:-form-item-description"
      >
        Your best password
      </p>
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 mt-2 w-full"
      data-testid="sign-in-button"
      type="submit"
    >
      Sign In
    </button>
  </form>
</div>
`;
